document/create.go

```diff
@@ -75,90 +75,61 @@ func NewFromStruct(s interface{}) (Document, error) {
 		return nil, errors.New("expected struct or pointer to struct")
 	}
 
-	return structDocument{ref: ref}, nil
+	return newFromStruct(ref)
 }
 
-type structDocument struct {
-	ref reflect.Value
-}
-
-var _ Document = (*structDocument)(nil)
-
-func (s structDocument) Iterate(fn func(field string, value Value) error) error {
-	l := s.ref.NumField()
-
-	tp := s.ref.Type()
+func newFromStruct(ref reflect.Value) (Document, error) {
+	var fb FieldBuffer
+	l := ref.NumField()
+	tp := ref.Type()
 
 	for i := 0; i < l; i++ {
-		sf := tp.Field(i)
-		if sf.PkgPath != "" {
+		f := reflect.Indirect(ref.Field(i))
+		if !f.IsValid() {
 			continue
 		}
 
-		var name string
-		if gtag, ok := sf.Tag.Lookup("genji"); ok {
-			if gtag == "-" {
-				continue
-			}
-
-			name = gtag
-		} else {
-			name = strings.ToLower(sf.Name)
-		}
-
-		f := s.ref.Field(i)
+		sf := tp.Field(i)
 
-		v, err := NewValue(f.Interface())
-		if err != nil {
-			if err.(*ErrUnsupportedType) != nil {
-				continue
+		switch f.Kind() {
+		case reflect.Struct:
+			d, err := newFromStruct(f)
+			if err != nil {
+				return nil, err
 			}
-			return err
-		}
-
-		err = fn(name, v)
-		if err != nil {
-			return err
-		}
-	}
-
-	return nil
-}
-
-func (s structDocument) GetByField(field string) (Value, error) {
-	tp := s.ref.Type()
 
-	var sf reflect.StructField
-	var ok bool
-
-	ln := tp.NumField()
-	for i := 0; i < ln; i++ {
-		sf = tp.Field(i)
-		if gtag, found := sf.Tag.Lookup("genji"); found && gtag == field {
-			ok = true
-			break
-		}
-		if strings.ToLower(sf.Name) == field {
-			ok = true
-			break
+			// Embedded struct.
+			if sf.Anonymous {
+				err := d.Iterate(func(field string, value Value) error {
+					fb.Add(field, value)
+					return nil
+				})
+
+				if err != nil {
+					return nil, err
+				}
+			} else {
+				fb.Add(strings.ToLower(sf.Name), NewDocumentValue(d))
+			}
+		default:
+			// Ignore unexported fields.
+			if sf.PkgPath == "" {
```

Comment:
Try to always get rid of the edge case first:

```go
if sf.PkgPath != "" {
  return &fb, nil
}

...
```

https://medium.com/@matryer/line-of-sight-in-code-186dd7cdea88

Reply:
Yes with `continue` we just ignore the unexported field.